### ISSUE TITLE
Question on programming style

### ISSUE THREAD

**Issue by ColinEberhardt** [xtuc/js-webassembly-interpreter#56]
Looking at the implementation of `add`, there is quite a bit of repetition:

    case 'add': {

      switch (instruction.object) {

      case 'i32': {
        const [c1, c2] = pop2('i32', 'i32');

        pushResult(
          binopi32(c2, c1, '+')
        );

        break;
      }

      case 'i64': {
        const [c1, c2] = pop2('i64', 'i64');

        pushResult(
          binopi64(c2, c1, '+')
        );

        break;
      }

      case 'f32': {
        const [c1, c2] = pop2('f32', 'f32');

        pushResult(
          binopf32(c2, c1, '+')
        );

        break;
      }

      case 'f64': {
        const [c1, c2] = pop2('f64', 'f64');

        pushResult(
          binopf64(c2, c1, '+')
        );

        break;
      }

      default:
        throw new RuntimeError(
          'Unsupported operation ' + instruction.id + ' on ' + instruction.object
        );

      }

      break;
    }

The following refactor gives the same functionality (ignoring the unsupported operation path for now):

    case 'add': {

      const binopComputer = {
        'i32': binopi32,
        'i64': binopi64,
        'f32': binopf32,
        'f64': binopf64,
      };

      const [c1, c2] = pop2(instruction.object, instruction.object);

      pushResult(
        binopComputer[instruction.object](c2, c1, '+')
      );

      break;

    }

Is a more 'long hand' approach being used for reasons of performance? Or is it worth exploring refactoring further?

**Comment by maurobringolf** [comment]
I doubt that this refactoring has a measurable performance impact, so from my perspective this refactoring would definitely be a win 👍

**Comment by ColinEberhardt** [comment]
Cool, I’ll propose some ‘light touch’ refactoring in a PR before I do more work on the binary ops.

**Comment by xtuc** [comment]
It's definitely more readable.

FYI I tried to split the kernel/exec file into a more convenient way to declare operations, the PR is here https://github.com/xtuc/js-webassembly-interpreter/pull/18. Unfortunately it became much more slower (https://github.com/xtuc/js-webassembly-interpreter/commit/1fdad4f43a073a2b28cfae58ce9a9619bd67e61f), and I had to revert it.

We have a make target if you want to run the benchmark, I believe it will be fine because it's not the whole interpreter like I tried.
